### PR TITLE
Fix bug that can't send file.

### PR DESCRIPTION
**Pull request by su-chang**
Fix the bug that can't send file. #91

### CODE REVIEW

File: src/puppet-puppeteer.ts [hunk 1]
```diff
@@ -1427,51 +1427,89 @@ export class PuppetPuppeteer extends Puppet {
...
+      } catch (e) {
+        log.error('PuppetPuppeteer', 'uploadMedia() uploadMedia exception: %s', e.message)
+        throw new Error('uploadMedia err: ' + e.message)
+      }
+    }
+    let funcList = []
+    for (let i = 0; i < bufferData.length - 1; i++) {
+      funcList.push(await getMediaId(bufferData[i], i))
     }
+    // use promise.all() make the former upload of this buffer quickly
```
Comment (windmemory) [inline src/puppet-puppeteer.ts]: You could do something like this here to get the last one:
```ts
const lastBuffer = bufferData.pop()
await Promise.all(bufferData.map(getMediaId))
const mediaId = await getMediaId(lastBuffer, bufferData.length)
```

File: src/puppet-puppeteer.ts [hunk 2]
```diff
@@ -1427,51 +1427,89 @@ export class PuppetPuppeteer extends Puppet {
...
-          reject(e)
-        }
-      })
-    } catch (e) {
-      log.error('PuppetPuppeteer', 'uploadMedia() uploadMedia exception: %s', e.message)
-      throw new Error('uploadMedia err: ' + e.message)
+      await Promise.all(funcList)
+      const lastOne = bufferData.length - 1
+      mediaId = await getMediaId(bufferData[lastOne], lastOne)
+    } catch {
```
Comment (windmemory) [inline src/puppet-puppeteer.ts]: I think we need to catch the error and log it here.

File: src/puppet-puppeteer.ts [hunk 3]
```diff
@@ -1427,51 +1427,84 @@ export class PuppetPuppeteer extends Puppet {
...
+     *  chunk: the index of chunks
+     */
+    const BASE_LENGTH = 512 * 1024
+    const chunks = Math.ceil(buffer.length / BASE_LENGTH)
+
+    const bufferData = []
+    for (let i = 0; i < chunks; i++) {
+      let tempBuffer = buffer.slice(i * BASE_LENGTH, (i + 1) * BASE_LENGTH)
+      bufferData.push(tempBuffer)
+    }
```
Comment (windmemory) [inline src/puppet-puppeteer.ts]: Let's do this like this:

```ts
const bufferData = [];
while (bufferData.length) {
  bufferData.push(buffer.splice(0, BASE_LENGTH)
}
```

File: src/puppet-puppeteer.ts (hunk 2 above)
Comment (su-chang) [reply src/puppet-puppeteer.ts]: Done

File: src/puppet-puppeteer.ts (hunk 1 above)
Comment (su-chang) [reply src/puppet-puppeteer.ts]: Done

**Comment by su-chang** [comment]
@huan 
I have found the change `1` to `pic`, `2` to `video`, `3` to `doc` relate to this commit. =>[Commit Detail Code](https://github.com/Chatie/webwx-app-tracker/commit/e91e284e446aac22883c28f4a4bfcab7f674ec5f#diff-80c9da80c8ca080b37c6c49864880646R1107).

It means that:
```javascript
r.MSGTYPE_IMAGE: n = "pic";
r.MSGTYPE_VIDEO: n = "video";
default: n = "doc"
```
and then 

`n` will assign to `mediatype`.

About `puppeteer`, our `WebMediaType` is also be used to assign `mediatype` for upload media.

This commit be committed on 19 Nov 2018, but the problem `cant send file in puppeteer` appearance these days. so it makes me confuse.

BTW, I find `split the uploading file into chunks` when I find issue about `can't send file`,  this PR also resolved this issue #12.

Thx for ur reading.

**Comment by windmemory** [comment]
> This commit be committed on 19 Nov 2018, but the problem cant send file in puppeteer appearance these days. so it makes me confuse.

I think the verification is on the server side, and during the debug phase, we saw that using different `mediaType` will result in different `mediaId`. The way of uploading the file is not the problem, the point is what `mediaId` is allowed to call the `sendMedia` endpoint. So I guess is that, WeChat updated their way of sending media file, but their api still allow the new `mediaId` and the old one. And recently, they add the validation for different `mediaId`.

**Comment by huan** [comment]
@su-chang Thank you very much for the additional information and they are very valuable.

I'll merge this PR for now, and it would be fantastic if we could refactor it to use the `factory` code inside puppeteer.

Good work!

**Comment by huan** [comment]
This PR should be published on v0.17.23 or later version.

**Pull request merged**